bpe.go

```diff
@@ -0,0 +1,110 @@
+package bpe
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+
+	"github.com/sirupsen/logrus"
+)
+
+type TokenId int32
+
+type DecodedToken []TokenId
+
+type Rule struct {
+	left   TokenId
+	right  TokenId
+	result TokenId
+}
+
+type SpecialTokens struct {
+	unk TokenId
+	pad TokenId
+	bos TokenId
+	eos TokenId
+}
+
+type Model struct {
+	char2id       map[rune]TokenId
+	id2char       map[TokenId]rune
+	rules         []Rule
+	recipe        map[TokenId]DecodedToken
+	revRecipe     map[string]TokenId
+	specialTokens SpecialTokens
+}
+
+func NewModel(nRules int) *Model {
```

Comment:
```suggestion
func newModel(nRules int) *Model {
```

because we are not supposed to construct new models in the code, only load